src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs

```diff
@@ -3363,18 +3370,47 @@ public override string? Namespace
 
         #region Attributes
 
-        public override Guid GUID
+        public override unsafe Guid GUID
         {
             get
             {
-                Guid result = default;
-                GetGUID(ref result);
+                TypeHandle th = GetNativeTypeHandle();
+                if (th.IsTypeDesc || th.AsMethodTable()->IsArray)
+                {
+                    return Guid.Empty;
+                }
+
+                Guid result;
+#if FEATURE_COMINTEROP
+                if (IsCOMObject)
```

Comment:
```suggestion
                if (TypeHandle.AreSameType(th, TypeHandle.TypeHandleOf<__ComObject>()))
```
Lower-level alternative for `IsGenericCOMObjectImpl()`